### PR TITLE
Switch calculation engine to use strongly typed values

### PR DESCRIPTION
**Pull request by jahav**
Obviously WIP for #1812 , but reasonably feature complete. Needs to be
* parted into far smaller, sensible commits. It's very unpleasant to read through such a big chunk of code.
* do benchmarks
* return again DateTime/TimeSpan from cell if cell has correct data type
* make SignatureAdapter more presentable, though it is a next phase, not this commit.
* review and review again, 

tl;dr;
* Use visitor to (I am pretty sure array formulas and dynamic array formulas will need a different evaluation visitor)
* Use strong types for values instead of object
* Use excel semantic for operators
* haven't modified function (except ABS, HYPERLINK, SUBTOTAL - demo + just easier than writing adapters)
* Errors are now enum (though function) and are returned from cell.Value instead of exceptions.
* prepare support for arrays

# OneOf
I really like the idea, it is just so easy to forget something and with the OneOf approach, it can't happen. Unfortunately, it has some drawbacks:
* structures are large, e.g. AnyValue is 8 bytes for synblk, 8 bytes for TypeHandle, 4 for index, 1 for logical, 8 for double, 8 for reference to a string, 4 for enum (could be changed to 1 byte), 8 for reference and 8 for array. 39 bytes + 16 overhead. I would like to replace the cell value that is always parsed with the AnyValue in the future and it is just too large and no way to decrease the size
* lambdas have closures in most cases. There is no way to pass a value to the each lambda, so the references value creates a closure on each call. Not only is it annoying because of allocations, JIT won't inline functions with closure (at least based on what I read).
* OneOf structures aren't readonly = defensive copies everywhere.
* All operators were in a static extension class
* package not signed (though that is pretty minor)

So in the end I made a bare-bone version version, with features we need.

# CalcContext
We need a context in many places, i.e. what sheet and cell is formula being evaluated (pretty much anything with references). I tried to keep various methods on worksheet/workbook backward compatible and only throw if we really need the context.

# Precedents are broken
It passes the tests, but any function returning a reference will break it (e.g. `A1:IF(Other!Z100, C100, Z 50)`)

### CODE REVIEW

**Comment by jahav** [comment]
@Pankraty Could you please take a look (not review, just look) at few files: CalcContext, AnyValue/ScalarValue, CalculationVisitor if it is in line with the CalcEngine redesign we were talking about?

**Comment by jahav** [comment]
I made a quick and simple benchmark against 0.96. Evaluation seems to be about the same, but parsing has taken a significant decrease with XLParser.

* SingleCellCalculationEvaluation - evaluation speed of `1 * 2 / (\"7\" + 8 - (1 + 6 * ( 2 * 3)))`
* SingleCellFunctionEvaluation  - evaluation speed of `SUM(1,2,3, IF(4 = 4, 5, 6))`
* SingleCellRangeEvaluation - evaluation speed of `SUM(A1:A100)` where `A1:A100` has value of 1
* DoubleCellRangeEvaluation  - evaluation speed of `SUM(A101:Z200)` where `A101:Z200` has a value `SUM(1, 2)`
* SingleCellCalculationParsing - parsing speed of `1 * 2 / (\"7\" + 8 - (1 + 6 * ( 2 * 3)))`
* SingleCellFunctionParsing - parsing speed of `SUM(1,2,3, IF(4 = 4, 5, 6))`

Benchmark gist - https://gist.github.com/jahav/3e244376d4a87e46782d74416e7fc011

issue-1812-formula-types branch (has XLParser and new type system)
``` ini

BenchmarkDotNet=v0.13.1, OS=Windows 10.0.22000
AMD Ryzen 5 5500U with Radeon Graphics, 1 CPU, 12 logical and 6 physical cores
.NET SDK=6.0.302
  [Host]     : .NET Core 3.1.27 (CoreCLR 4.700.22.30802, CoreFX 4.700.22.31504), X64 RyuJIT
  DefaultJob : .NET Core 3.1.27 (CoreCLR 4.700.22.30802, CoreFX 4.700.22.31504), X64 RyuJIT


```
|                          Method |         Mean |       Error |      StdDev |
|-------------------------------- |-------------:|------------:|------------:|
| SingleCellCalculationEvaluation |     121.0 ns |     2.10 ns |     2.06 ns |
|    SingleCellFunctionEvaluation |     115.9 ns |     0.79 ns |     0.74 ns |
|       SingleCellRangeEvaluation |     120.4 ns |     0.70 ns |     0.66 ns |
|       DoubleCellRangeEvaluation |     119.9 ns |     0.72 ns |     0.64 ns |
|    SingleCellCalculationParsing | 272,656.1 ns | 1,260.17 ns | 1,052.30 ns |
|       SingleCellFunctionParsing | 223,074.3 ns |   950.97 ns |   889.54 ns |


Version 0.96
``` ini

BenchmarkDotNet=v0.13.1, OS=Windows 10.0.22000
AMD Ryzen 5 5500U with Radeon Graphics, 1 CPU, 12 logical and 6 physical cores
.NET SDK=6.0.302
  [Host]     : .NET Core 3.1.27 (CoreCLR 4.700.22.30802, CoreFX 4.700.22.31504), X64 RyuJIT
  DefaultJob : .NET Core 3.1.27 (CoreCLR 4.700.22.30802, CoreFX 4.700.22.31504), X64 RyuJIT


```
|                          Method |       Mean |    Error |   StdDev |
|-------------------------------- |-----------:|---------:|---------:|
| SingleCellCalculationEvaluation |   119.6 ns |  0.23 ns |  0.20 ns |
|    SingleCellFunctionEvaluation |   120.9 ns |  0.43 ns |  0.41 ns |
|       SingleCellRangeEvaluation |   120.2 ns |  0.47 ns |  0.44 ns |
|       DoubleCellRangeEvaluation |   112.9 ns |  1.50 ns |  1.40 ns |
|    SingleCellCalculationParsing | 2,227.7 ns |  5.73 ns |  5.08 ns |
|       SingleCellFunctionParsing | 3,701.1 ns | 12.83 ns | 11.38 ns |

**Comment by Pankraty** [comment]
@jahav I started looking at the PR, will give my feedback a bit later. 
May I ask you to rerun the benchmarks with `MemoryDiagnoser`? I have a gut feeling that xlparser may be allocating heap memory as hell. If so, on environments with limited memory (containers, lambdas) this can slow down parsing even worse die to frequent GC.

**Comment by jahav** [comment]
@Pankraty I 've rerun the benchmark with the `[MemoryDiagnoser]` as you asked and I realized that I screwed up. The original benchmark didn't invalidate formulas. 
Considering numbers I got and I don't think it is worthwhile for you to look at the branch for now.

# Current develop branch
<html>
<body>
<pre style="color: rgb(0, 0, 0); font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><code>BenchmarkDotNet=v0.13.1, OS=Windows 10.0.22000
AMD Ryzen 5 5500U with Radeon Graphics, 1 CPU, 12 logical and 6 physical cores
.NET SDK=6.0.302
  [Host]     : .NET Core 3.1.27 (CoreCLR 4.700.22.30802, CoreFX 4.700.22.31504), X64 RyuJIT
  Job-JPRWPH : .NET Core 3.1.27 (CoreCLR 4.700.22.30802, CoreFX 4.700.22.31504), X64 RyuJIT
</code></pre><pre style="color: rgb(0, 0, 0); font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><code>InvocationCount=1  UnrollFactor=1  
</code></pre>

Method | Mean | Error | StdDev | Median | Allocated
-- | -- | -- | -- | -- | --
SingleCellCalculationEvaluation | 39.23 μs | 3.878 μs | 10.615 μs | 37.20 μs | 2 KB
SingleCellFunctionEvaluation | 59.86 μs | 5.506 μs | 15.348 μs | 54.60 μs | 3 KB
SingleCellRangeEvaluation | 4,943.63 μs | 143.662 μs | 409.875 μs | 4,849.55 μs | 1,830 KB
DoubleCellRangeEvaluation | 7,266.02 μs | 270.811 μs | 798.494 μs | 7,056.20 μs | 2,015 KB
SingleCellCalculationParsing | 16.26 μs | 1.204 μs | 3.317 μs | 15.55 μs | 2 KB
SingleCellFunctionParsing | 37.79 μs | 2.692 μs | 7.594 μs | 35.20 μs | 3 KB


</body>
</html>

# Current issue-1812-formula-types branch
<html>
<body>
<pre style="color: rgb(0, 0, 0); font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><code><br class="Apple-interchange-newline">
BenchmarkDotNet=v0.13.1, OS=Windows 10.0.22000
AMD Ryzen 5 5500U with Radeon Graphics, 1 CPU, 12 logical and 6 physical cores
.NET SDK=6.0.302
  [Host]     : .NET Core 3.1.27 (CoreCLR 4.700.22.30802, CoreFX 4.700.22.31504), X64 RyuJIT
  Job-OMUDGD : .NET Core 3.1.27 (CoreCLR 4.700.22.30802, CoreFX 4.700.22.31504), X64 RyuJIT
</code></pre><pre style="color: rgb(0, 0, 0); font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><code>InvocationCount=1  UnrollFactor=1  
</code></pre>

Method | Mean | Error | StdDev | Median | Gen 0 | Allocated
-- | -- | -- | -- | -- | -- | --
SingleCellCalculationEvaluation | 367.3 μs | 5.65 μs | 6.94 μs | 366.3 μs | - | 19 KB
SingleCellFunctionEvaluation | 415.6 μs | 33.74 μs | 95.71 μs | 378.1 μs | - | 21 KB
SingleCellRangeEvaluation | 5,554.6 μs | 226.97 μs | 662.09 μs | 5,486.4 μs | - | 1,836 KB
DoubleCellRangeEvaluation | 12,705.7 μs | 528.77 μs | 1,525.63 μs | 11,970.4 μs | 2000.0000 | 4,881 KB
SingleCellCalculationParsing | 345.7 μs | 6.86 μs | 17.33 μs | 339.4 μs | - | 17 KB
SingleCellFunctionParsing | 315.9 μs | 12.95 μs | 36.52 μs | 299.2 μs | - | 18 KB


</body>
</html>

**Comment by jahav** [comment]
I declare it good enough. There are some potential problems but in most cases are dependent on some other code I don't want to fix at this PR. At some point, I have to stop. This PR is already massive.

I have learned a lot about profiling, that is good.

Main changes since last time:
* Added blank to type system. I tried to 
* Streamline AST - merged `EmptyArgumentNode` and `ErrorNode` to a `ScalarNode`

Key findings for future PR:
Perf findings:
* *PERF* Cache doesn't work, like at all. It uses WeakReference, but there is no other place where the reference is kept. Consequence - formula is parsed far more often than it should because it is removed from cache.
* *PERF* - Need a visitor to pre-calculate reference for A1:B1 type areas. Currenly always calculated during each evaluation.
* *MEMORY* - Most allocations happen during boxing in `XLCell.RecalculateFormula` or `CalcEngine.Evaluate`. That is not really that big of a problem, it is collected quickly, but replace with a struct arguments and properties in the cell (internal only, keep API same)
* *BUG* - determining precedent cells for a formula is non-working. In the future, create CalcChains and multiple workers.
* *BUG* - `XLNamedRange` assumes that all things behind name is a list of ranges. That is not true, name is pretty much only alias for a formula (e.g. =0.5).

# Benchmark
Last benchmark - here is a [fixed gist](https://gist.github.com/jahav/5716ab8cd590bc05d42d1991b7e231a7). I forgot to require cells to be recalculated in my first benchmarks.

It is kind of unfair, because I am using a new SUM function that actually works as it should (though not optimized, just straightforward implementation), but original was just select all things to memory and sum it. Legacy functions don't have high priority for me - pretty much all will be replaced because of bugs or performance reasons.

*TL;DR;*: Simple formula is slower, more complex formulas are faster and require less memory. There is no reason to case after some super optimization, when a get property accessor takes up 1% of time, we have passed a point when it is worth optimizing the code and we should optimize algorithms (multiple workers ect.). What was intended here was to provide an bedrock for easy implementation of functions that work the same way as excel does.

Speed-Sum-Getter-Takes-Percent

## Current version
``` ini

BenchmarkDotNet=v0.13.2, OS=Windows 11 (10.0.22000.856/21H2)
AMD Ryzen 5 5500U with Radeon Graphics, 1 CPU, 12 logical and 6 physical cores
.NET SDK=6.0.302
  [Host]     : .NET Core 3.1.27 (CoreCLR 4.700.22.30802, CoreFX 4.700.22.31504), X64 RyuJIT AVX2
  DefaultJob : .NET Core 3.1.27 (CoreCLR 4.700.22.30802, CoreFX 4.700.22.31504), X64 RyuJIT AVX2


```
|                          Method |         Mean |      Error |     StdDev |     Gen0 | Allocated |
|-------------------------------- |-------------:|-----------:|-----------:|---------:|----------:|
| SingleCellCalculationEvaluation |     2.496 μs |  0.0170 μs |  0.0142 μs |   0.0763 |     160 B |
|    SingleCellFunctionEvaluation |     2.826 μs |  0.0169 μs |  0.0150 μs |   0.6599 |    1384 B |
|       SingleCellRangeEvaluation |   718.942 μs |  3.6413 μs |  3.0406 μs |  30.2734 |   64797 B |
|       DoubleCellRangeEvaluation | 4,642.513 μs | 19.4962 μs | 16.2802 μs | 273.4375 |  586237 B |

## Original version

``` ini

BenchmarkDotNet=v0.13.2, OS=Windows 11 (10.0.22000.856/21H2)
AMD Ryzen 5 5500U with Radeon Graphics, 1 CPU, 12 logical and 6 physical cores
.NET SDK=6.0.302
  [Host]     : .NET Core 3.1.27 (CoreCLR 4.700.22.30802, CoreFX 4.700.22.31504), X64 RyuJIT AVX2
  DefaultJob : .NET Core 3.1.27 (CoreCLR 4.700.22.30802, CoreFX 4.700.22.31504), X64 RyuJIT AVX2


```
|                          Method |           Mean |         Error |        StdDev |     Gen0 |     Gen1 |     Gen2 | Allocated |
|-------------------------------- |---------------:|--------------:|--------------:|---------:|---------:|---------:|----------:|
| SingleCellCalculationEvaluation |       957.3 ns |       9.69 ns |       9.07 ns |   0.0153 |        - |        - |      32 B |
|    SingleCellFunctionEvaluation |       983.4 ns |       7.21 ns |       6.39 ns |   0.0839 |   0.0019 |        - |     176 B |
|       SingleCellRangeEvaluation | 4,635,247.4 ns |  63,228.57 ns |  59,144.04 ns | 539.0625 | 257.8125 | 148.4375 | 1871002 B |
|       DoubleCellRangeEvaluation | 6,816,119.8 ns | 116,950.83 ns | 134,680.74 ns | 671.8750 | 265.6250 | 140.6250 | 2060662 B |
